### PR TITLE
Admin stats

### PR DESCRIPTION
**Pull request by TimCsaky**
# Description

reference: [ticket 856](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-856)
I've finally added a few extra details to the admin 'service client table'.
details to show:
- the dates that app was promoted to each environment
- users in the team for each app
- full name for the application

note: This is just a first draft. I will be adding catches, fixing my bad syntax, adding some unit tests etc tomorrow. I'm more keen to make sure that the data is coming from the right place.
The dates of promotion are from the lifecycle table which I know we will plan to change later.

I did want to get all teh data in one go and re-purpose it for other metrics (passed down to child components) but that hasnt happenned yet.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/frontend/src/components/admin/Clients.vue [hunk 1]
```diff
@@ -76,19 +122,17 @@ export default {
...
-          const clients = Object.keys(reduced).map(k => {
+          const data = response.data;
+          const clients = Object.keys(data).map(k => {
+            let client = data[k];
             return {
-              name: k,
-              dev: reduced[k].includes('dev'),
-              test: reduced[k].includes('test'),
-              prod: reduced[k].includes('prod')
+              acronym: client.acronym,
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/Clients.vue:129]: i've changeed the structure of the data from the backend so this mapping to data table is different

File: app/frontend/src/components/admin/Clients.vue [hunk 2]
```diff
@@ -12,20 +14,62 @@
...
+            <table class="kc-nested-table">
+              <!-- environments -->
+              <tr>
+                <td width="40px"></td>
+                <td>
+                  <strong>created:</strong>
+                </td>
+                <td>{{ (item.environments.DEV) ? new Date(item.environments.DEV.created).toLocaleDateString() : '' }}</td>
+                <td>{{ (item.environments.TEST) ? new Date(item.environments.TEST.created).toLocaleDateString() : '' }}</td>
+                <td>{{ (item.environments.PROD) ? new Date(item.environments.PROD.created).toLocaleDateString() : '' }}</td>
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/Clients.vue:45]: I know this is not pretty

File: app/frontend/src/components/admin/Clients.vue [hunk 3]
```diff
@@ -106,10 +150,44 @@ export default {
...
+}
+.kc-table >>> tr.v-data-table__expanded.v-data-table__expanded__content {
+  -webkit-box-shadow: none !important;
+  box-shadow: none !important;
+}
+.kc-table tr.last td {
+  border-bottom: 1px !important;
+  padding-bottom: 10px;
+}
+.kc-nested-table {
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/Clients.vue:184]: i have some layout issues complicated by vuetify.
i may need to re-think the layout of the expanded row. will check with Mina

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-169

**Comment by loneil** [comment]
Dates aren't rendering and alignment is borked
image

File: app/frontend/src/components/admin/Clients.vue [hunk 4]
```diff
@@ -12,20 +14,62 @@
...
+        :expanded.sync="expanded"
+        item-key="acronym"
+        show-expand
       >
+        <template v-slot:expanded-item="{ headers, item }">
+          <td :colspan="headers.length">
+            <table class="kc-nested-table">
+              <!-- environments -->
+              <tr>
+                <td width="40px"></td>
```
Comment (loneil) [inline app/frontend/src/components/admin/Clients.vue:39]: I'm not sure using a table inside this table row is going to work out (I might be wrong...). Especially with set pixel widths, as the column sizes for the parent table will vary with screen size. See my attached screenshot in a comment on this pr.

Comment (TimCsaky) [reply app/frontend/src/components/admin/Clients.vue:39]: yeah. i'll do something with divs

**Comment by TimCsaky** [comment]
> 
> 
> Dates aren't rendering and alignment is borked
> image

i believe only the WORG and TEMP clients will have data in the database. I dont know what's up with the invalid date yet

File: app/src/routes/v1/keycloak.js [hunk 5]
```diff
@@ -8,28 +8,88 @@ const Problem = require('api-problem');
...
+      const acronymnDetailsFromDb = await acronyms.getAcronym(obj.acronym);
+
+      // if a corresponding acronym in db (there wont be in local dev enviroment)
+      if (acronymnDetailsFromDb) {
+
+        obj.acronymnDetails = acronymnDetailsFromDb.dataValues;
+
+        // promotions (from lifecycle table for now)
+        const promotions = await lifecycleService.findLatestPromotions(obj.acronymnDetails.acronymId);
+        promotions.forEach(function (k) {
```
Comment (jujaga) [inline app/src/routes/v1/keycloak.js:71]: This can be an arrow function as there isn't any context sensitive binds.
```
promotions.forEach(function (k) {
```

File: app/src/routes/v1/keycloak.js [hunk 6]
```diff
@@ -8,28 +8,88 @@ const Problem = require('api-problem');
...
+          obj.environments[k.dataValues.env].created = k.dataValues.createdAt;
+          obj.environments[k.dataValues.env].updated = k.dataValues.updatedAt;
+        });
+
+        // users
+        const usersList = await acronyms.getUsers(obj.acronym);
+        obj.users = usersList;
```
Comment (jujaga) [inline app/src/routes/v1/keycloak.js:78]: This can be collapsed into a one line await assignment.
```
obj users = await acronyms.getUsers(obj.acronym);
```

**Comment by loneil** [comment]
Users will have to be vertical or some other layout as well
image

File: app/src/routes/v1/keycloak.js [hunk 7]
```diff
@@ -8,28 +8,88 @@ const Problem = require('api-problem');
...
+
+  getDetails(reduced)
+    .then(data => {
+      if (data === null) {
+        return new Problem(404).send(res);
+      } else {
+        res.status(200).json(data);
+      }
+    });
 });
```
Comment (jujaga) [inline app/src/routes/v1/keycloak.js:93]: There is a lot of new business logic added here in the router layer. This logic should really be in its own component as we want the routing layer to focus on figuring out what needs to be achieved to satisfy an incoming request, whereas a component focuses on what logic needs to be achieved in order to provide the callee with the information they need.

**Comment by jujaga** [comment]
There's a bit of formatting strangeness happening on the users list/array.
table formatting

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-169

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        loading-text="Loading... Please wait"
+        :expanded.sync="expanded"
+        item-key="acronym"
+        show-expand
+
+      >
+        <!-- environments -->
+        <template v-slot:item.dev="{item}">
+          <div v-if="item.dev">
+            <v-tooltip top>
```
Comment (TimCsaky) [inline app/frontend/src/components/admin/ServiceClients.vue:37]: promotion date shows in tooltip. would be nice to have this tool tip appear on click event as well for mobile users but i dont think that's working

**Comment by TimCsaky** [comment]
obviously only the clients stored in the pr's database will show the extended details (TEMP and WORK i think)

File: app/frontend/src/components/admin/ServiceClients.vue (hunk 8 above)
Comment (jujaga) [reply app/frontend/src/components/admin/ServiceClients.vue:36]: You can probably add some toggle logic by binding visibility to a v-model. Caveat is when doing this, I suspect the default hover event will no longer be listened to, so you'd have to re-implement that behavior in addition to the on-click toggle.
Reference: https://vuetifyjs.com/en/components/tooltips/#visibility

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            </div>
+
+            <div class="kc-expanded">
+              <!-- users -->
+              <strong>Users: </strong>
+              <ul>
+                <li
+                  v-for="(user, index) in item.users"
+                  :key="index"
+                >{{ user.user.firstName + ' ' + user.user.lastName + ((item.users.length > (index + 1)) ? ',' : '') }}</li>
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:88]: You probably don't need that many parentheses for evaluating where in the array it is - this expression would be equivalent and suffice:
`(index <= item.users.length) ? ',' : ''`

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+              </template>
+              <span>Updated: {{ formatEnvDate(item.environments.PROD) }}</span>
+            </v-tooltip>
+          </div>
+        </template>
+
+        <!-- expanded row -->
+        <template v-slot:expanded-item="{ headers, item }">
+          <td :class="[responsiveCell]"></td>
+          <td :colspan="headers.length - 1" :class="[responsiveCell]">
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:68]: You'd have to double check, but seeing as you want the responsiveCell behavior on all table cells, you may be able to just pull that class out into its own div as a parent of it all.
```
<div :class="[responsiveCell]">
  <td ...
  ...
</div>
```

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+    };
+  },
+  methods: {
+
+    formatEnvDate(env) {
+      return (env && env.created) ?  this.formatDate(env.created): '-';
+    },
+    formatDate(date){
+      return new Date(date).toLocaleString();
+    },
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:134]: These two functions can probably be chained together into one - this is likely why the tooltip and created date outputs are inconsistent as you're calling them both differently.

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+              acronymDetails: client.acronymDetails,
+              dev: client.environments.DEV,
+              test: client.environments.TEST,
+              prod: client.environments.PROD,
+              environments: client.environments,
+              users: client.users
+            };
+          });
+
+          if (clients.length == 0) {
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:155]:
```
if (!clients.length) {
```

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+            this.showTableAlert('info', 'No Service Clients found');
+          }
+          this.serviceClients = clients;
+        })
+        .catch(() => {
+          this.showTableAlert('error', 'No response from server');
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:161]: Please add a //TODO note regarding global error notification behavior as this will be relevant if/when we tackle that refactor.

File: app/frontend/src/components/admin/ServiceClients.vue (hunk 9 above)
Comment (TimCsaky) [reply app/frontend/src/components/admin/ServiceClients.vue:88]: Your code wouldnt work for showing commas.
this would work:
`( (index + 1)  < item.users.length ) ? ',' : ''
`
but its basically what i have

File: app/frontend/src/components/admin/ServiceClients.vue (hunk 10 above)
Comment (TimCsaky) [reply app/frontend/src/components/admin/ServiceClients.vue:68]: not sure i understand. the class '.v-data-table__mobile-table-row' needs to be applied to all <td> elements on small screens. thats how the vuetify datatable columns are responsive

File: app/frontend/src/components/admin/ServiceClients.vue (hunk 9 above)
Comment (jujaga) [reply app/frontend/src/components/admin/ServiceClients.vue:88]: Let's just run with that then. Either way, generally it's better to do conditionals with the first term being the value to check, and the second term being the constant you're comparing against (which is what your latest example has. I'd still remove the inner-most parentheses though:
```
( index + 1 < item.users.length ) ? ',' : ''
```

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 14]
```diff
@@ -127,10 +127,11 @@ export default {
   methods: {
 
     formatEnvDate(env) {
-      return (env && env.created) ?  this.formatDate(env.created): '-';
+      return (env && env.created) ?  this.formatDate(env.created): 'N/A';
     },
     formatDate(date){
-      return new Date(date).toLocaleString();
+
```
Comment (jujaga) [inline app/frontend/src/components/admin/ServiceClients.vue:133]: Unnecessary whitespace line

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-k8vopl-tools.pathfinder.gov.bc.ca/dashboard?id=nr-get-token-pr-169

File: app/src/components/clients.js [hunk 15]
```diff
@@ -1,18 +1,95 @@
...
+
+      // app details
+      const acronymObj = await acronyms.getAcronym(sc.acronym);
+
+      // if a corresponding acronym in db
+      if (acronymObj) {
+        sc.acronymDetails = acronymObj.dataValues;
+
+        // promotions (from lifecycle table for now)
+        const promotions = await lifecycleService.findLatestPromotions(sc.acronymDetails.acronymId);
```
Comment (loneil) [inline app/src/components/clients.js:29]: Discussed refactoring this so we're not doing db calls per iteration in the loop. Can live with it for now but let's do a backlog ticket to track a performance enhancement. Might not ever be important but if this ever gets used outside of the admin context, or starts getting slow with rising numbers of clients we'll have to address.

File: app/src/components/clients.js [hunk 16]
```diff
@@ -1,18 +1,95 @@
...
+
+        promotions.forEach(promotion => {
+          if (promotion.length) {
+            const data = promotion[0].dataValues;
+            sc.environments[data.env].created = data.createdAt;
+            sc.environments[data.env].updated = data.updatedAt;
+          }
+        });
+
+        sc.users = await acronyms.getUsers(sc.acronym);
```
Comment (loneil) [inline app/src/components/clients.js:39]: Same as above, db calls in loop.

File: app/src/services/lifecycle.js [hunk 17]
```diff
@@ -33,4 +33,39 @@ module.exports = {
...
+   * @returns {object} A lifecycle object if it exists
+   */
+  async getLatestEnvAction(acronymId, env) {
+    const lifecycle = await db.Lifecycle.findAll({
+      where: {
+        acronymId: acronymId,
+        env: env
+      },
+      order: [
+        ['updatedAt', 'DESC']
```
Comment (loneil) [inline app/src/services/lifecycle.js:51]: This will work I think (since in this case updatedAt will always be the same), but createdAt would be the relevant field here if I'm not mistaken?

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+          <td :colspan="headers.length - 1" :class="[responsiveCell]">
+            <div class="kc-expanded">
+              <!-- app details -->
+              <strong>Name: </strong>
+              <span>{{ item.name }}</span>
+            </div>
+
+            <div class="kc-expanded">
+              <!-- app details -->
+              <strong>Acronymn Created: </strong>
```
Comment (loneil) [inline app/frontend/src/components/admin/ServiceClients.vue:77]: typo

File: app/src/services/lifecycle.js (hunk 17 above)
Comment (TimCsaky) [reply app/src/services/lifecycle.js:51]: I guess so. especially if a password reset would match a record.
i can change that now to created.

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+  display: none !important;
+}
+tr.v-data-table__expanded__content
+  td.v-data-table__mobile-table-row:not(:nth-child(1)) {
+  padding: 0;
+}
+td.v-data-table__mobile-table-row div.kc-expanded {
+  padding: 0.2rem 1rem;
+}
+</style>
```
[the remark this answers is not in the document]
Comment (TimCsaky) [reply app/frontend/src/components/admin/ServiceClients.vue:233]: I was thinking the same thing but read something about sass not handling >>> properly. I'll look into that again later.

File: app/src/components/clients.js (hunk 15 above)
Comment (TimCsaky) [reply app/src/components/clients.js:29]: thanks for pointing this out. i should have really caught that.

File: app/src/services/lifecycle.js (hunk 17 above)
Comment (TimCsaky) [reply app/src/services/lifecycle.js:51]: yeah i assumed either was good. we're calling it 'environment last updated' in the ui.

File: app/frontend/src/components/admin/ServiceClients.vue [hunk 20]
```diff
@@ -22,7 +22,7 @@
         class="kc-table"
         :headers="headers"
         :items="serviceClients"
-        :items-per-page="10"
+        :items-per-page="7"
```
Comment (loneil) [inline app/frontend/src/components/admin/ServiceClients.vue:25]: This change doesn't seem to be working. I think it only takes 5, 10, 15...
10 is better IMO too

Comment (TimCsaky) [reply app/frontend/src/components/admin/ServiceClients.vue:25]: ok. sounds good.

Comment (TimCsaky) [reply app/frontend/src/components/admin/ServiceClients.vue:25]: i know you suggested making them alphabetical too. but i didnt get that done yet

Comment (jujaga) [reply app/frontend/src/components/admin/ServiceClients.vue]: Let's leave the alphabetical stuff out for now.
